### PR TITLE
[gnn-survey]graph isomorphism network

### PR DESCRIPTION
**Pull request by joshchang1112**

### CODE REVIEW

File: research/gnn-survey/README.md [hunk 1]
```diff
@@ -88,10 +88,27 @@ python train.py \
   --sparse_features=True
 ```
 
+*   GIN results on cora (81.7% average test accuracy):
+
+`python train.py \
```
Comment (ppham27) [inline research/gnn-survey/README.md:93]: Enclose in three backticks so it's formatted properly

File: research/gnn-survey/layers.py [hunk 2]
```diff
@@ -167,12 +167,14 @@ def call(self, inputs):
...
 
-    attn = tf.sparse.add(adj * attn_row,
+    attn = tf.sparse.add(adj * attn_row, 
                          adj * tf.transpose(attn_col, perm=[1, 0]))
     attn = tf.sparse.SparseTensor(
         indices=attn.indices,
         values=self.leakyrelu(attn.values),
         dense_shape=attn.dense_shape)
+
+    attn = tf.sparse.reorder(attn)
```
Comment (ppham27) [inline research/gnn-survey/layers.py:177]: This shouldn't be necessary since the previous operation doesn't manipulate the indices?

File: research/gnn-survey/layers.py [hunk 3]
```diff
@@ -183,3 +185,74 @@ def call(self, inputs):
...
+  def __init__(self, mlp_layers, output_dim, dropout_rate, learn_eps=False, sparse=True, **kwargs):
+    """Initializes the GraphIsomorphismLayer.
+
+    Args:
+      mlp layers: (int) Number of mlp layers
+      output_dim: (int) Output dimension of gcn layer
+      dropout_rate: (float) Dropout probability
+      learn_eps: (bool) Whether to learn the epsilon weighting
+      sparse: (bool) sparse: Whether features are sparse
+      **kwargs: Keyword arguments for tf.keras.layers.Layer.
```
Comment (ppham27) [inline research/gnn-survey/layers.py:202]: Enclose in backticks.

File: research/gnn-survey/layers.py [hunk 4]
```diff
@@ -183,3 +185,74 @@ def call(self, inputs):
...
+
+
+class GraphIsomorphismLayer(tf.keras.layers.Layer):
+  """Single graph isomorphism layer."""
+
+  def __init__(self, mlp_layers, output_dim, dropout_rate, learn_eps=False, sparse=True, **kwargs):
+    """Initializes the GraphIsomorphismLayer.
+
+    Args:
+      mlp layers: (int) Number of mlp layers
```
Comment (ppham27) [inline research/gnn-survey/layers.py:197]: Missing underscore

File: research/gnn-survey/layers.py [hunk 5]
```diff
@@ -183,3 +185,74 @@ def call(self, inputs):
...
+              shape=(self.output_dim, self.output_dim),
+              initializer='random_normal',
+              trainable=True)
+        )
+
+    if self.learn_eps:
+      self.eps = self.add_weight(
+          name='bias',
+          shape=(1,),
+          initializer='random_normal',
```
Comment (ppham27) [inline research/gnn-survey/layers.py:242]: You could set trainable=self.learn_eps. Usually bias is initialized to zero: https://github.com/tensorflow/tensorflow/blob/v2.3.1/tensorflow/python/keras/layers/core.py#L1136.

File: research/gnn-survey/layers.py [hunk 6]
```diff
@@ -183,3 +185,74 @@ def call(self, inputs):
...
+              shape=(input_shape[0][-1], self.output_dim),
+              initializer='random_normal',
+              trainable=True)
+          )
+      else:
+        self.weight.append(
+          self.add_weight(
+              name='weight'+str(i+1),
+              shape=(self.output_dim, self.output_dim),
+              initializer='random_normal',
```
Comment (ppham27) [inline research/gnn-survey/layers.py:234]: It's not clear that you actually want random_normal here. glorot_uniform or glorot_normal may be preferable since they scale by the shape dimensions, which is why it's the Keras default: https://www.tensorflow.org/api_docs/python/tf/keras/layers/Dense

File: research/gnn-survey/models.py [hunk 7]
```diff
@@ -12,9 +12,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 """Modeling for GNNs."""
-from layers import GraphAttnLayer
+from layers import GraphAttnLayer, SparseGraphAttnLayer
 from layers import GraphConvLayer
-from layers import SparseGraphAttnLayer
```
Comment (ppham27) [inline research/gnn-survey/models.py:17]: Keep this as a separate line.

File: research/gnn-survey/models.py [hunk 8]
```diff
@@ -200,3 +195,96 @@ def call(self, inputs):
...
+
+
+class GINBlock(tf.keras.layers.Layer):
+  """Graph isomorphism block."""
+
+  def __init__(self, mlp_layers, hidden_dim, dropout_rate, learn_eps, sparse, **kwargs):
+    """Initializes a GIN block.
+
+    Args:
+      mlp layers: (int) Number of mlp layers in GIN
```
Comment (ppham27) [inline research/gnn-survey/models.py:207]: Missing underscore.

**Review by ppham27 — CHANGES_REQUESTED** [review]
Just some small details to satisfy our internal linters.

File: research/gnn-survey/README.md (hunk 1 above)
Comment (joshchang1112) [reply research/gnn-survey/README.md:93]: OK!

File: research/gnn-survey/layers.py (hunk 2 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:177]: I'm not very sure but I think this may help sparse.softmax function below apply correctly. Since the indices have the wrong dimension at first and the adj matrix is not symmetric, I think that it may be added?(Let softmax apply on write dimension, and acc also improve a lot with reorder function)

File: research/gnn-survey/layers.py (hunk 4 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:197]: OK!

File: research/gnn-survey/layers.py (hunk 3 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:202]: Do you mean that enclose `tf.keras.layers.Layer` in backticks?

File: research/gnn-survey/models.py (hunk 7 above)
Comment (joshchang1112) [reply research/gnn-survey/models.py:17]: OK!

File: research/gnn-survey/models.py (hunk 8 above)
Comment (joshchang1112) [reply research/gnn-survey/models.py:207]: OK!

File: research/gnn-survey/layers.py (hunk 5 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:242]: OK

File: research/gnn-survey/layers.py (hunk 6 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:234]: OK! Should I replace all weight to glorot_normoal initializer? or just GIN layer need to replace? 
Thanks:)

**Comment by joshchang1112** [comment]
I just modified the inconsistent coding style and layers building function. 
I also have some question mentioned above~ 
Thanks for your review:)

File: research/gnn-survey/layers.py (hunk 2 above)
Comment (ppham27) [reply research/gnn-survey/layers.py:178]: Okay, add just add a comment that says that, then.

File: research/gnn-survey/layers.py (hunk 3 above)
Comment (ppham27) [reply research/gnn-survey/layers.py]: Yes, thanks.

File: research/gnn-survey/layers.py (hunk 6 above)
Comment (ppham27) [reply research/gnn-survey/layers.py]: If you can replace all of them, that would be great.

Comment (joshchang1112) [reply research/gnn-survey/layers.py]: OK:)

File: research/gnn-survey/layers.py (hunk 2 above)
Comment (joshchang1112) [reply research/gnn-survey/layers.py:178]: OK!
